libs/pdi-scanner/src/rust/protocol/image.rs

```diff
@@ -1,9 +1,42 @@
-use image::{DynamicImage, GrayImage};
+use rayon::{prelude::ParallelIterator, slice::ParallelSlice};
+
+use image::GrayImage;
+
+use crate::client::ImageCalibrationTables;
 
 use super::types::ScanSideMode;
 
 pub const DEFAULT_IMAGE_WIDTH: u32 = 1728;
 
+/// Applies image calibration to a single row of pixel data based on the
+/// provided white and black calibration tables (retrieved from the scanner).
+/// The formula used is based on guidance from PDI.
+fn apply_image_calibration(
+    row: &[u8],
+    white_calibration_table: &[u8],
+    black_calibration_table: &[u8],
+) -> Vec<u8> {
+    assert!(
+        row.len() == white_calibration_table.len() && row.len() == black_calibration_table.len(),
+        "Image calibration tables must be the same length as the row"
+    );
+
+    row.iter()
+        .enumerate()
+        .map(|(index, &pixel)| {
+            let white_calibration = white_calibration_table[index];
+            let black_calibration = black_calibration_table[index];
```

Comment:
FWIW, I did some benchmarking and didn't see any difference with this and the `.zip().zip()` approach 🤷

Reply:
Ok, cool, I'll switch to that